### ISSUE TITLE
Duplicate State Confusion

### ISSUE THREAD

**Issue by RickeyWard** [wbkd/react-flow#777]
In all the examples it starts with 
`const [elements, setElements] = useState(initialelements);`
then that is passed into `ReactFlow`. alterations such as deletes and connections in the examples all edit this state object but the control itself keeps track of the elements in another state object as part of the internal store and you have to monitor that state for changes with `const elements = useStoreState(store => store.elements);` then you are exposed ` const setElements = useStoreActions((action) => action.setElements);` to modify that store but nothing bubbles that action up to your own store.

So how is this supposed to be used in a real app? Do you never use useStoreActions' setElements? or do you somehow exclusively use it?

If you have a node with variable handles, having a button in that node to allow you delete a handle should delete all the edges associated with it, no problem, use setElements from useStoreActions... except when you do it's only gone until something edits your top level useState, and then it pops back because it was only gone from the internal state. How do you keep this in sync? Use a 2nd context with your own useState? what's the point of the internal one then? I've had to put listeners on everything just to extract the state when something changes because otherwise it will get lost. But you can't listen to the internal store in the same component that hosts the ReactFlow component because you get into a render loop.

does anyone have an suggestions on how this is intended to work?

**Comment by moklick** [comment]
Hey @RickeyWard,

I understand that the state handling is not 100% clear, because some information about that is missing in our documentation. 

To your questions: 

TLDR: Create your own store/state with elements, only alter that state, never use the internal `setElements` ;)

React Flow handles some state internally like dragging and selecting of nodes and edges in order to keep it simple. Otherwise you would need to implement a handler for all of these actions by yourself. This already leads to an async state of your elements and the elements inside React Flow. The position don't match anymore. But that should be no problem because you can alter the position from outside to set a new position or receive the actual position by using `toObject` or `elements` of the internal state. However when you alter elements like adding, removing or updating you should use your own state where all the information about the elements are. We are building a relatively complex app with React Flow and we never use the internal React Flow `setElements` within our app but our own `setElements` that alter our app state. Does this help you?

**Comment by RickeyWard** [comment]
Thanks for the insight.

I will attempt never using `useStoreActions((action) => action.setElements)`  for state management and create my own context object to get my root setElements into my nodes.
